### PR TITLE
Users/tihuang/newartifact

### PR DESCRIPTION
**Pull request by TingluoHuang**

### CODE REVIEW

**Comment by TingluoHuang** [comment]
Still needs to update all doc links within task.json.

File: Tasks/CopyFiles/CopyFiles.ts [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+// contents is a multiline input containing glob patterns
+var contents: string[] = tl.getDelimitedInput('Contents', '\n');
+var sourceFolder: string = tl.getPathInput('SourceFolder');
+var targetFolder: string = tl.getPathInput('TargetFolder');
+
+var cleanTargetFolderString: string = tl.getInput('CleanTargetFolder');
+var overWriteString: string = tl.getInput('OverWrite');
+var useCommonRootString: string = tl.getInput('UseCommonRoot');
+
+var cleanTargetFolder: boolean = cleanTargetFolderString.trim().toLocaleLowerCase() == "true";
```
Comment (TingluoHuang) [inline Tasks/CopyFiles/CopyFiles.ts]: ===

Comment (TingluoHuang) [inline Tasks/CopyFiles/CopyFiles.ts]: use "===".

File: Tasks/CopyFiles/CopyFiles.ts [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+var allFiles: string[] = [];
+
+// remove folder path
+for (var i: number = 0; i < allPaths.length; i++) {
+    if (!fs.statSync(allPaths[i]).isDirectory()) {
+        allFiles.push(allPaths[i]);
+    }
+}
+
+if (includeContents && allFiles) {
```
Comment (TingluoHuang) [inline Tasks/CopyFiles/CopyFiles.ts]: if(includeContents&&allfiles&&includeContents.length>0 &&allfiles.length>0)

File: Tasks/CopyFiles/CopyFiles.ts [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+            var targetDir = path.dirname(targetPath);
+                            
+            if (!createdFolders[targetDir]) {
+                tl.debug("Creating folder " + targetDir);
+                tl.mkdirP(targetDir);
+                createdFolders[targetDir] = true;
+            }
+            
+            var exists = fs.existsSync(targetPath),
+                stats = exists && fs.statSync(targetPath);
```
Comment (TingluoHuang) [inline Tasks/CopyFiles/CopyFiles.ts]: var exists = fs.existsSync(targetPath);
var stats = exists && fs.statSync(targetPath);

File: Tasks/CopyFiles/task.json [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+      "defaultValue": "false",
+      "required": false,
+      "helpMarkDown": "Relace existing file in target folder",
+      "groupName":"advanced"
+    },
+    {
+      "name": "UseCommonRoot",
+      "type": "boolean",
+      "label": "Use Common Root",
+      "defaultValue": "false",
```
Comment (TingluoHuang) [inline Tasks/CopyFiles/task.json]: default to true?

File: Tasks/CopyPublishBuildArtifacts/task.json [hunk 5]
```diff
@@ -0,0 +1,80 @@
+﻿{
+  "id": "1D341BB0-2106-458C-8422-D00BCEA6512A",
+  "name": "CopyPublishBuildArtifacts",
```
Comment (TingluoHuang) [inline Tasks/CopyPublishBuildArtifacts/task.json]: How we want to name this task?

File: Tasks/PublishBuildArtifacts/task.json [hunk 6]
```diff
@@ -12,27 +12,19 @@
...
+    "Patch": 0
   },
   "demands": [
   ],
-    "minimumAgentVersion": "1.83.0",
+    "minimumAgentVersion": "1.91.0",
   "inputs": [
     {
-      "name": "CopyRoot",
+      "name": "Content",
```
Comment (TingluoHuang) [inline Tasks/PublishBuildArtifacts/task.json]: how do we want to name this input parameter?

File: Tasks/CopyFiles/task.json (hunk 4 above)
Comment (chrisrpatterson) [inline Tasks/CopyFiles/task.json]: What does UseCommoneRoot do?

File: Tasks/PublishBuildArtifacts/task.json (hunk 6 above)
Comment (chrisrpatterson) [inline Tasks/PublishBuildArtifacts/task.json]: Call it ArtifactContents and make the description Folder containing files that you want to include in the artifact

File: Tasks/CopyFiles/task.json (hunk 4 above)
Comment (TingluoHuang) [inline Tasks/CopyFiles/task.json]: Short the path on target folder by remove the common root of files need copied.
Ex.
source:  c:\a\b\c\test1.txt
              c:\a\b\c\test2.txt
              c:\a\b\c\test3.txt
SourceFolder = c:\a
TargetFolder = d:\t
With using common root.
copy result:
            d:\t\test1.txt
            d:\t\test2.txt
            d:\t\test3.txt
Otherwise:
            d:\t\b\c\test1.txt
            d:\t\b\c\test2.txt
            d:\t\b\c\test3.txt

Comment (chrisrpatterson) [inline Tasks/CopyFiles/task.json]: You should write up a document on how to use this task as I think it could be confusing for people.

Comment (bryanmacfarlane) [inline Tasks/CopyFiles/task.json]: No we shouldn't default to true.

IMHO, common root seems very fragile.  Other things that react post copy suddenly break as other changes are made.

Is it better to have an advanced optional argument which is the common root? c:\a\b\c
After all, I might want c:\a\b and end up with
d:\t\c\test*.txt in the target

Seems more deterministic, less spooky, and as an optional advanced option, more control

Thoughts?

File: Tasks/CopyFiles/task.json [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+      "helpMarkDown": "Delete all existing files in target folder before copy.",
+      "groupName":"advanced"
+    },
+    {
+      "name": "OverWrite",
+      "type": "boolean",
+      "label": "Over Write",
+      "defaultValue": "false",
+      "required": false,
+      "helpMarkDown": "Relace existing file in target folder",
```
Comment (bryanmacfarlane) [inline Tasks/CopyFiles/task.json]: typo Replace

File: Tasks/CopyFiles/task.json (hunk 4 above)
Comment (bryanmacfarlane) [inline Tasks/CopyFiles/task.json]: Actually, I think since this is a new task, we should start minimal and add as we're sure and it's solid.  With that line of thinking, we should remove the common root.  Leave the implementation in the code but hard code to false and for now don't expose.  

Once this is out there, we can't take it back.

Comment (chrisrpatterson) [inline Tasks/CopyFiles/task.json]: I think what we are really trying to accomplish here is to flatten the folder hierarchy and maybe that should be the option.

I am not sure I like the idea of having the user specify the common root because I think it is hard to explain and specify.  We already have enough issues explaining the copy patterns in the existing task and how those apply to the Copy Root.

Comment (chrisrpatterson) [inline Tasks/CopyFiles/task.json]: I think that makes the most sense.

File: Tasks/CopyFiles/task.json [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+      "groupName":"advanced"
+    },
+    {
+      "name": "UseCommonRoot",
+      "type": "boolean",
+      "label": "Use Common Root",
+      "defaultValue": "false",
+      "required": false,
+      "helpMarkDown": "Shorter the path in target folder by using common root of copied files",
```
Comment (bryanmacfarlane) [inline Tasks/CopyFiles/task.json]: I'm advocating removing but if we keep it should be Shorten

File: Tasks/CopyFiles/task.json [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+      "name": "CleanTargetFolder",
+      "type": "boolean",
+      "label": "Clean Target Folder",
+      "defaultValue": "false",
+      "required": false,
+      "helpMarkDown": "Delete all existing files in target folder before copy.",
+      "groupName":"advanced"
+    },
+    {
+      "name": "OverWrite",
```
Comment (bryanmacfarlane) [inline Tasks/CopyFiles/task.json]: Is overwrite only relevant if we don't clean target?

The reason I ask is there's the case within one match/copy invocation where a file is matched multiple times.  In other words I don't think one instance of a task should *ever* copy the same file twice in the same invocation regardless of this setting.

File: Tasks/CopyFiles/task.json (hunk 8 above)
Comment (chrisrpatterson) [inline Tasks/CopyFiles/task.json]: Let’s remove it for the first version.

File: Tasks/CopyFiles/task.json [hunk 10]
```diff
@@ -0,0 +1,87 @@
+﻿{
+  "id": "5BFB729A-A7C8-4A78-A7C3-8D717BB7C13C",
+  "name": "CopyFiles",
+  "friendlyName": "Copy Files",
+  "description": "Copy files from source folder to target folder using minimatch patterns",
+  "helpMarkDown": "[More Information](http://go.microsoft.com/fwlink/?LinkID=)",
```
Comment (bryanmacfarlane) [inline Tasks/CopyFiles/task.json]: Make the fwlink point to the same as the current copyPublish combo task we have now.  In that documentation, we can cover, copy, publish, copyPublish

File: Tasks/CopyPublishBuildArtifacts/task.json [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+  "friendlyName": "Copy Publish Build Artifacts",
```
Comment (bryanmacfarlane) [inline Tasks/CopyPublishBuildArtifacts/task.json]: Copy and Publish Artifacts

File: Tasks/CopyPublishBuildArtifacts/task.json [hunk 12]
```diff
@@ -0,0 +1,80 @@
+﻿{
```
Comment (bryanmacfarlane) [inline Tasks/CopyPublishBuildArtifacts/task.json]: Why are we doing an add here of an existing task?  We're losing all history.

We should do git mv PublishBuildArtifacts CopyPublishBuildArtifacts and then edit json properties

File: Tasks/PublishBuildArtifacts/Strings/resources.resjson/en-US/resources.resjson [hunk 13]
```diff
@@ -6,10 +6,8 @@
   "loc.helpMarkDown": "[More Information](http://go.microsoft.com/fwlink/?LinkID=613725)",
   "loc.description": "Publish Build artifacts to the server or a file share using minimatch patterns",
   "loc.instanceNameFormat": "Publish Artifact: $(ArtifactName)",
```
Comment (bryanmacfarlane) [inline Tasks/PublishBuildArtifacts/Strings/resources.resjson/en-US/resources.resjson]: If you do a git mv as suggested above then the new task in the same path space will be a git add.  I think that will reflect what really happened.

File: Tasks/PublishBuildArtifacts/task.json [hunk 14]
```diff
@@ -65,15 +57,9 @@
...
-    "PowerShell": {
-      "target": "$(currentDirectory)\\PublishBuildArtifacts.ps1",
-      "argumentFormat": "",
-      "workingDirectory": "$(currentDirectory)",
-      "platforms" : ["windows"]
-    },
+  "execution": {    
     "Node": {
-      "target": "publishBuildArtifacts.js",
+      "target": "PublishBuildArtifacts.js",
```
Comment (bryanmacfarlane) [inline Tasks/PublishBuildArtifacts/task.json]: Why did you change this?  it's actually publishBuildArtifacts and *nix is case sensitive

File: Tasks/CopyFiles/task.json (hunk 9 above)
Comment (TingluoHuang) [inline Tasks/CopyFiles/task.json]: We won't copy same file multiple times within the task. but there is case that user can have multiple copyfile tasks within their definition.
